server/cluster/store_limiter.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License
+
+package cluster
+
+import (
+	"sync"
+
+	"github.com/pingcap/kvproto/pkg/pdpb"
+	"github.com/pingcap/log"
+	"github.com/pingcap/pd/server/schedule"
+	"go.uber.org/zap"
+)
+
+// StoreLimiter adjust the store limit dynamically
+type StoreLimiter struct {
+	m       sync.RWMutex
+	oc      *schedule.OperatorController
+	scene   *schedule.StoreLimitScene
+	state   *State
+	current LoadState
+}
+
+// NewStoreLimiter builds a store limiter object using the operator controller
+func NewStoreLimiter(c *schedule.OperatorController) *StoreLimiter {
+	return &StoreLimiter{
+		oc:      c,
+		state:   NewState(),
+		scene:   schedule.DefaultStoreLimitScene(),
+		current: LoadStateNone,
+	}
+}
+
+// Collect the store statistics and update the cluster state
+func (s *StoreLimiter) Collect(stats *pdpb.StoreStats) {
+	s.m.Lock()
+	defer s.m.Unlock()
+
+	log.Debug("collected statistics", zap.Reflect("stats", stats))
+	s.state.Collect((*StatEntry)(stats))
+
+	rate := float64(0)
+	state := s.state.State()
+	switch state {
+	case LoadStateIdle:
```

Comment:
It would be more smooth with more states, it is hard to say how many states should be used. I think    it is better to keep the design without special reasons.